<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Differential Gene Expression Analysis across Cell Clusters — DEA • scrabble</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Differential Gene Expression Analysis across Cell Clusters — DEA" />
<meta property="og:description" content="This function returns the genes differentially expressed in a given cluster of cells relative to the remaining cells in the matrix. If a list of clusters is provided, DEA will be computed for each cluster in turn. Differential expression is calculated according to fold changes and p-values. The user can specify the fold change and p-value with which genes are deemed (DE-)significant. With default arguments, genes are considered significant if FC &amp;gt;= 2 and p-value
(Benjamin-Hochberg-adjusted) &amp;lt;= 0.01. Also with default arguments, fold-change values are returned and genes are sorted by fold-change (largest first). The user can instead ask for genes to be sorted by p-value (most significant first) or for p-values to be returned or both. Alternatively, if the user requires fold change values and p-values for all genes (not just those that are significant), they can set return.full = TRUE. This parameter will return a list containing (1) all clusters, (2) fold-change values for all genes, (3) p-values for all genes, (4) the gene names and indices of the DE-significant genes and (5-7) the fold change and p-value cutoffs used as well as the method used to adjust p-values for multiple testing." />
<meta name="twitter:card" content="summary" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">scrabble</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/scrabble.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Differential Gene Expression Analysis across Cell Clusters</h1>
    
    <div class="hidden name"><code>DEA.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function returns the genes differentially expressed in a given cluster of cells relative to the remaining cells in the matrix. If a list of clusters is provided, DEA will be computed for each cluster in turn. Differential expression is calculated according to fold changes and p-values. The user can specify the fold change and p-value with which genes are deemed (DE-)significant. With default arguments, genes are considered significant if FC &gt;= 2 and p-value
(Benjamin-Hochberg-adjusted) &lt;= 0.01. Also with default arguments, fold-change values are returned and genes are sorted by fold-change (largest first). The user can instead ask for genes to be sorted by p-value (most significant first) or for p-values to be returned or both. Alternatively, if the user requires fold change values and p-values for all genes (not just those that are significant), they can set return.full = TRUE. This parameter will return a list containing (1) all clusters, (2) fold-change values for all genes, (3) p-values for all genes, (4) the gene names and indices of the DE-significant genes and (5-7) the fold change and p-value cutoffs used as well as the method used to adjust p-values for multiple testing.</p>
    </div>

    <pre class="usage"><span class='fu'>DEA</span>(<span class='no'>m</span>, <span class='no'>clusters</span>, <span class='kw'>FC</span> <span class='kw'>=</span> <span class='fl'>2</span>, <span class='kw'>is.log</span> <span class='kw'>=</span> <span class='no'>T</span>, <span class='kw'>p</span> <span class='kw'>=</span> <span class='fl'>0.01</span>,
  <span class='kw'>adjust.method</span> <span class='kw'>=</span> <span class='st'>"BH"</span>, <span class='kw'>sort</span> <span class='kw'>=</span> <span class='no'>T</span>, <span class='kw'>p.sort</span> <span class='kw'>=</span> <span class='no'>F</span>, <span class='kw'>return.full</span> <span class='kw'>=</span> <span class='no'>F</span>,
  <span class='kw'>return.p</span> <span class='kw'>=</span> <span class='no'>F</span>, <span class='kw'>fast</span> <span class='kw'>=</span> <span class='no'>T</span>)</pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>m</th>
      <td><p>matrix</p></td>
    </tr>
    <tr>
      <th>clusters</th>
      <td><p>List of cluster memberships (character vectors). If provided, will be used instead of computed.. Default: NULL</p></td>
    </tr>
    <tr>
      <th>FC</th>
      <td><p>fold change cutoff. Default: 2</p></td>
    </tr>
    <tr>
      <th>is.log</th>
      <td><p>values are in log2 (used in fold_changes). Default: T</p></td>
    </tr>
    <tr>
      <th>p</th>
      <td><p>p-value cutoff. Default: 0.01</p></td>
    </tr>
    <tr>
      <th>adjust.method</th>
      <td><p>Correction for multiple tests. If not desired, set to 'none'. Default: 'BH'</p></td>
    </tr>
    <tr>
      <th>sort</th>
      <td><p>returning values are sorted by fold change (largest first). Default: T</p></td>
    </tr>
    <tr>
      <th>p.sort</th>
      <td><p>returning values are sorted by p-value (most significant first). Default: F</p></td>
    </tr>
    <tr>
      <th>return.full</th>
      <td><p>return cluster membership and cluster DE genes. Default: F</p></td>
    </tr>
    <tr>
      <th>return.p</th>
      <td><p>return cluster DE p-values. Default: F</p></td>
    </tr>
    <tr>
      <th>fast</th>
      <td><p>if TRUE, computes p-values for significant fold-change values only. Default: T</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>with default parameters, will return a list of differentially expressed genes in statistically significant clusters. Differentially expressed genes are those passing fold change and p-value cutoffs. Statistically significant clusters are those passing nsig1 and nsig2 cutoffs. Within clusters, gene values displayed are fold changes and genes are ordered by highest FC first.</p>
    <h2 class="hasAnchor" id="details"><a class="anchor" href="#details"></a>Details</h2>

    <p>DETAILS</p>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      <li><a href="#value">Value</a></li>
      <li><a href="#details">Details</a></li>
    </ul>

  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Julie Laffy.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.9000.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


